File: src/main/java/twao/assigners/Assigner.java
package twao.assigners;

import twao.villages.AllyVillage;
import twao.villages.TargetVillage;
import twao.villages.Village;

import java.util.Comparator;
import java.util.List;

class Assigner implements Runnable {

    final List<TargetVillage> targets;
    final List<AllyVillage> attackingVillages;

    final Village relativityPoint;
    final boolean assigningFakes;
    private final AssigningStrategy assigningStrategy;

    Assigner(List<TargetVillage> targets, List<AllyVillage> attackingVillages, Village relativityPoint, boolean assigningFakes,
                    AssigningStrategy assigningStrategy) {
        this.targets = targets;
        this.attackingVillages = attackingVillages;

        this.relativityPoint = relativityPoint;
        this.assigningFakes = assigningFakes;
        this.assigningStrategy = assigningStrategy;
    }

    public void run() {
        assigningStrategy.run(this);
    }

    static void sort(List<? extends Village> villages, Village relativityPoint) {
        for (Village vil : villages) {
            vil.setRelativeDistance(relativityPoint);
        }

        villages.sort(Comparator.comparing(Village::getRelativeDistance));
    }
}

File: src/main/java/twao/assigners/Assigner.kt
package twao.assigners

import twao.Player
import twao.VillageAssignment
import twao.villages.AllyVillage
import twao.villages.TargetVillage
import twao.villages.Village
import java.util.*

abstract class Assigner internal constructor(
        protected val targets: MutableList<TargetVillage>,
        protected val resources: MutableList<AllyVillage>,
        protected val mainReferencePoint: Village,
        protected val isAssigningFakes: Boolean
) : Runnable {

    protected abstract val targetsQueue: PriorityQueue< Pair<TargetVillage, Int> >
    protected abstract val resourcesQueue: PriorityQueue< Pair<AllyVillage, Int> >

    protected abstract val offAction: Player.(VillageAssignment) -> Unit
    protected abstract val fakeAction: Player.(VillageAssignment) -> Unit

    protected open fun putResourcesToQueue(referencePoint: Village) {
        resources.forEach { resourcesQueue.offer( Pair(it, Village.distance(it, referencePoint)) ) }
    }

    protected open fun putTargetsToQueue(referencePoint: Village) {
        targets.forEach { targetsQueue.offer( Pair(it, Village.distance(it, referencePoint)) ) }
    }

    protected fun assign(allyVillage: AllyVillage, targetVillage: TargetVillage, distance: Int) {
        val assignment = VillageAssignment(
                departure = allyVillage,
                destination = targetVillage,
                squaredDistance = distance
        )

        when (isAssigningFakes) {
            true -> allyVillage.owner.fakeAction(assignment)
            false -> allyVillage.owner.offAction(assignment)
        }
    }
}